diff --git a/config.def.h b/config.def.h
index 1edb647..4f3de64 100644
--- a/config.def.h
+++ b/config.def.h
@@ -1,20 +1,33 @@
 /* See LICENSE file for copyright and license details. */
 /* Default settings; can be overriden by command line. */
 
-static int topbar = 1;                      /* -b  option; if 0, dmenu appears at bottom     */
+static int topbar      = 1;      /* -b  option; if 0, dmenu appears at bottom     */
+static int user_bh     = 24;     /* bar height, default 18 */
+static int fuzzy       = 1;      /* -F  option; if 0, dmenu doesn't use fuzzy matching     */
+static int colorprompt = 1;      /* -p  option; if 1, prompt uses SchemeSel, otherwise SchemeNorm */
+static int centered    = 0;      /* -c option; centers dmenu on screen */
+static int min_width   = 500;    /* minimum width when centered */
 /* -fn option overrides fonts[0]; default X11 font or font set */
 static const char *fonts[] = {
-	"monospace:size=10"
+	"JetBrainsMono Nerd Font Bandit:style=Medium:pixelsize=15",
+        "Twemoji:pixelsize=15:antialias=true:autohint=true"
 };
-static const char *prompt      = NULL;      /* -p  option; prompt to the left of input field */
+        /* "JoyPixels:pixelsize=15:antialias=true:autohint=true", */
+	/* "octicons:style=Medium:pixelsize=15", */
+	/* "Font Awesome 5 Free:style=Regular:size=8", */
+	/* "Font Awesome 5 Brands:style=Regular:size=8", */
+static const char *prompt   = NULL;      /* -p  option; prompt to the left of input field */
+static const char *symbol_1 = "";
+static const char *symbol_2 = "";
 static const char *colors[SchemeLast][2] = {
-	/*     fg         bg       */
-	[SchemeNorm] = { "#bbbbbb", "#222222" },
-	[SchemeSel] = { "#eeeeee", "#005577" },
-	[SchemeOut] = { "#000000", "#00ffff" },
+	            /*      fg         bg     */
+	[SchemeNorm] = { "#bbbbbb", "#1e1e1e" },
+	[SchemeSel]  = { "#eeeeee", "#005f87" },
+	[SchemeOut]  = { "#000000", "#00ffff" },
 };
-/* -l option; if nonzero, dmenu uses vertical list with given number of lines */
+/* -l and -g options; controls number of lines and columns in grid if > 0 */
 static unsigned int lines      = 0;
+static unsigned int columns    = 0;
 
 /*
  * Characters not considered part of a word while deleting words
diff --git a/config.mk b/config.mk
index 0df3fc8..6204f8a 100644
--- a/config.mk
+++ b/config.mk
@@ -5,8 +5,8 @@ VERSION = 5.1
 PREFIX = /usr/local
 MANPREFIX = $(PREFIX)/share/man
 
-X11INC = /usr/X11R6/include
-X11LIB = /usr/X11R6/lib
+X11INC = /usr/local/include
+X11LIB = /usr/local/lib
 
 # Xinerama, comment if you don't want it
 XINERAMALIBS  = -lXinerama
