diff --git a/dwm.c b/dwm.c
index 9fd0286..652d2e0 100644
--- a/dwm.c
+++ b/dwm.c
@@ -211,6 +211,7 @@ static void tag(const Arg *arg);
 static void tagmon(const Arg *arg);
 static void tile(Monitor *);
 static void togglebar(const Arg *arg);
+static void togglesystray();
 static void togglefloating(const Arg *arg);
 static void toggletag(const Arg *arg);
 static void toggleview(const Arg *arg);
@@ -1708,6 +1709,16 @@ togglebar(const Arg *arg)
 	arrange(selmon);
 }
 
+void
+togglesystray()
+{
+	if (showsystray)
+		XUnmapWindow(dpy, systray->win);
+	showsystray = !showsystray;
+	updatesystray();
+	updatestatus();
+}
+
 void
 togglefloating(const Arg *arg)
 {
