diff --git a/dwm.c b/dwm.c
index 664c527..c7bfe8e 100644
--- a/dwm.c
+++ b/dwm.c
@@ -178,6 +178,7 @@ static void grabkeys(void);
 static void incnmaster(const Arg *arg);
 static void keypress(XEvent *e);
 static void killclient(const Arg *arg);
+static void layoutmenu(const Arg *arg);
 static void manage(Window w, XWindowAttributes *wa);
 static void mappingnotify(XEvent *e);
 static void maprequest(XEvent *e);
@@ -1015,6 +1016,25 @@ killclient(const Arg *arg)
 	}
 }
 
+void
+layoutmenu(const Arg *arg)
+{
+	FILE *p;
+	char c[3], *s;
+	int i;
+
+	if (!(p = popen(layoutmenu_cmd, "r")))
+		 return;
+	s = fgets(c, sizeof(c), p);
+	pclose(p);
+
+	if (!s || *s == '\0' || c == '\0')
+		 return;
+
+	i = atoi(c);
+	setlayout(&((Arg) { .v = &layouts[i] }));
+}
+
 void
 manage(Window w, XWindowAttributes *wa)
 {
diff --git a/layoutmenu.sh b/layoutmenu.sh
new file mode 100644
index 0000000..a752556
--- /dev/null
+++ b/layoutmenu.sh
@@ -0,0 +1,13 @@
+#!/bin/sh
+
+cat <<EOF | xmenu
+[]= Tiledi                 0
+><> Floating               1
+[M] Monocle                2
+TTT Bstack                 3
+=== BstackH                4
+[@] Spiral                 5
+[\\] Dwindle               6
+|M| CenteredMaster         7
+>M> CenteredFloatingMaster 8
+EOF
