diff --git a/config.mk b/config.mk
index b6eb7e0..7084c33 100644
--- a/config.mk
+++ b/config.mk
@@ -1,5 +1,5 @@
 # dwm version
-VERSION = 6.3
+VERSION = 6.2
 
 # Customize below to fit your system
 
diff --git a/drw.c b/drw.c
index 4cdbcbe..d81c004 100644
--- a/drw.c
+++ b/drw.c
@@ -16,6 +16,9 @@ static const unsigned char utfmask[UTF_SIZ + 1] = {0xC0, 0x80, 0xE0, 0xF0, 0xF8}
 static const long utfmin[UTF_SIZ + 1] = {       0,    0,  0x80,  0x800,  0x10000};
 static const long utfmax[UTF_SIZ + 1] = {0x10FFFF, 0x7F, 0x7FF, 0xFFFF, 0x10FFFF};
 
+/* Show emoji color on status bar, 0 disable, 1 enable */
+static const int color_emoji = 1;
+
 static long
 utf8decodebyte(const char c, size_t *i)
 {
@@ -140,11 +143,13 @@ xfont_create(Drw *drw, const char *fontname, FcPattern *fontpattern)
 	 * https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=916349
 	 * and lots more all over the internet.
 	 */
-	FcBool iscol;
-	if(FcPatternGetBool(xfont->pattern, FC_COLOR, 0, &iscol) == FcResultMatch && iscol) {
-		XftFontClose(drw->dpy, xfont);
-		return NULL;
-	}
+    if (color_emoji == 0) { /* Config on drw.c line 19 */
+        FcBool iscol;
+        if(FcPatternGetBool(xfont->pattern, FC_COLOR, 0, &iscol) == FcResultMatch && iscol) {
+            XftFontClose(drw->dpy, xfont);
+            return NULL;
+        }
+    }
 
 	font = ecalloc(1, sizeof(Fnt));
 	font->xfont = xfont;
diff --git a/dwm.c b/dwm.c
index a96f33c..5e4d494 100644
--- a/dwm.c
+++ b/dwm.c
@@ -702,9 +702,6 @@ drawbar(Monitor *m)
 	unsigned int i, occ = 0, urg = 0;
 	Client *c;
 
-	if (!m->showbar)
-		return;
-
 	/* draw status first so it can be overdrawn by tags later */
 	if (m == selmon) { /* status is only drawn on selected monitor */
 		drw_setscheme(drw, scheme[SchemeNorm]);
