diff --git a/config.h b/config.h
index 8e5d90a..ffed145 100644
--- a/config.h
+++ b/config.h
@@ -3,17 +3,17 @@
 
 static int bottom = 0;
 static int embedded = 0;
-static int minpwlen = 32;
+static int minpwlen = 10;
 static int mon = -1;
 
-static const char *asterisk = "*";
+static const char *asterisk = "#";
 static const char *fonts[] = {
-	"monospace:size=10"
+	"JetBrainsMono Nerd Font Bandit:style=Medium,Regular:size=8"
 };
 static const char *prompt = NULL;
 static const char *colors[SchemeLast][4] = {
-	[SchemePrompt] = { "#bbbbbb", "#222222" },
-	[SchemeNormal] = { "#bbbbbb", "#222222" },
-	[SchemeSelect] = { "#eeeeee", "#005577" },
-	[SchemeDesc]   = { "#bbbbbb", "#222222" }
+	[SchemePrompt] = { "#eeeeee", "#005f87" },
+	[SchemeNormal] = { "#bbbbbb", "#1e1e1e" },
+	[SchemeSelect] = { "#eeeeee", "#005f87" },
+	[SchemeDesc]   = { "#eeeeee", "#005f87" }
 };
diff --git a/config.mk b/config.mk
index 4c71211..aef0ea1 100644
--- a/config.mk
+++ b/config.mk
@@ -7,8 +7,8 @@ BUGREPORT = https:\/\/github.com\/ritze\/pinentry-dmenu
 PREFIX    = /usr/local
 MANPREFIX = ${PREFIX}/share/man
 
-X11INC = /usr/X11R6/include
-X11LIB = /usr/X11R6/lib
+X11INC = /usr/local/include
+X11LIB = /usr/local/lib
 
 # Xinerama, comment if you don't want it
 XINERAMALIBS  = -lXinerama
