diff --git a/config.def.h b/config.def.h
index 9981ca3..8dd604d 100644
--- a/config.def.h
+++ b/config.def.h
@@ -48,10 +48,10 @@ static const bool ALPHA_LAYER = false;
 #ifdef _THUMBS_CONFIG
 
 /* thumbnail sizes in pixels (width == height): */
-static const int thumb_sizes[] = { 32, 64, 96, 128, 160 };
+static const int thumb_sizes[] = { 32, 64, 96, 128, 160, 360 };
 
 /* thumbnail size at startup, index into thumb_sizes[]: */
-static const int THUMB_SIZE = 3;
+static const int THUMB_SIZE = 5;
 
 #endif
 #ifdef _MAPPINGS_CONFIG
