diff --git a/config.def.h b/config.def.h
index 9981ca3..6ed5a87 100644
--- a/config.def.h
+++ b/config.def.h
@@ -48,10 +48,10 @@ static const bool ALPHA_LAYER = false;
 #ifdef _THUMBS_CONFIG
 
 /* thumbnail sizes in pixels (width == height): */
-static const int thumb_sizes[] = { 32, 64, 96, 128, 160 };
+static const int thumb_sizes[] = { 32, 64, 96, 128, 160, 360 };
 
 /* thumbnail size at startup, index into thumb_sizes[]: */
-static const int THUMB_SIZE = 3;
+static const int THUMB_SIZE = 4;
 
 #endif
 #ifdef _MAPPINGS_CONFIG
diff --git a/exec/key-handler b/exec/key-handler
index 3b50e4c..f6080c4 100755
--- a/exec/key-handler
+++ b/exec/key-handler
@@ -31,5 +31,6 @@ case "$1" in
 "C-comma")  rotate 270 ;;
 "C-period") rotate  90 ;;
 "C-slash")  rotate 180 ;;
+"C-d")      while read file; do mv "$file" ~/.trash; done ;;
 esac
 
