diff --git a/st.1 b/st.1
index 39120b4..57ddfb8 100644
--- a/st.1
+++ b/st.1
@@ -22,6 +22,8 @@ st \- simple terminal
 .IR line ]
 .RB [ \-w
 .IR windowid ]
+.RB [ \-z
+.IR fontsize ]
 .RB [[ \-e ]
 .IR command
 .RI [ arguments ...]]
@@ -44,6 +46,8 @@ st \- simple terminal
 .IR title ]
 .RB [ \-w
 .IR windowid ]
+.RB [ \-z
+.IR fontsize ]
 .RB \-l
 .IR line
 .RI [ stty_args ...]
@@ -91,6 +95,10 @@ defines the window title (default 'st').
 embeds st within the window identified by
 .I windowid
 .TP
+.BI \-z " fontsize"
+sets the default fontsize to
+.I fontsize
+.TP
 .BI \-l " line"
 use a tty
 .I line
diff --git a/st.c b/st.c
index a9338e1..ebdf360 100644
--- a/st.c
+++ b/st.c
@@ -793,15 +793,14 @@ ttynew(const char *line, char *cmd, const char *out, char **args)
 		break;
 	case 0:
 		close(iofd);
-		close(m);
 		setsid(); /* create a new process group */
 		dup2(s, 0);
 		dup2(s, 1);
 		dup2(s, 2);
 		if (ioctl(s, TIOCSCTTY, NULL) < 0)
 			die("ioctl TIOCSCTTY failed: %s\n", strerror(errno));
-		if (s > 2)
-			close(s);
+		close(s);
+		close(m);
 #ifdef __OpenBSD__
 		if (pledge("stdio getpw proc exec", NULL) == -1)
 			die("pledge\n");
diff --git a/x.c b/x.c
index 89786b8..75dd38c 100644
--- a/x.c
+++ b/x.c
@@ -4,6 +4,7 @@
 #include <limits.h>
 #include <locale.h>
 #include <signal.h>
+#include <stdlib.h>
 #include <sys/select.h>
 #include <time.h>
 #include <unistd.h>
@@ -1116,7 +1117,7 @@ xinit(int cols, int rows)
 		die("could not init fontconfig.\n");
 
 	usedfont = (opt_font == NULL)? font : opt_font;
-	xloadfonts(usedfont, 0);
+	xloadfonts(usedfont, defaultfontsize);
 
 	/* colors */
 	xw.cmap = XDefaultColormap(xw.dpy, xw.scr);
@@ -1588,9 +1589,8 @@ xseticontitle(char *p)
 	XTextProperty prop;
 	DEFAULT(p, opt_title);
 
-	if (Xutf8TextListToTextProperty(xw.dpy, &p, 1, XUTF8StringStyle,
-	                                &prop) != Success)
-		return;
+	Xutf8TextListToTextProperty(xw.dpy, &p, 1, XUTF8StringStyle,
+			&prop);
 	XSetWMIconName(xw.dpy, xw.win, &prop);
 	XSetTextProperty(xw.dpy, xw.win, &prop, xw.netwmiconname);
 	XFree(prop.value);
@@ -1602,9 +1602,8 @@ xsettitle(char *p)
 	XTextProperty prop;
 	DEFAULT(p, opt_title);
 
-	if (Xutf8TextListToTextProperty(xw.dpy, &p, 1, XUTF8StringStyle,
-	                                &prop) != Success)
-		return;
+	Xutf8TextListToTextProperty(xw.dpy, &p, 1, XUTF8StringStyle,
+			&prop);
 	XSetWMName(xw.dpy, xw.win, &prop);
 	XSetTextProperty(xw.dpy, xw.win, &prop, xw.netwmname);
 	XFree(prop.value);
@@ -2042,6 +2041,11 @@ main(int argc, char *argv[])
 	case 'v':
 		die("%s " VERSION "\n", argv0);
 		break;
+	case 'z':
+		defaultfontsize = strtod(EARGF(usage()), NULL);
+		if (!(defaultfontsize > 0))
+			usage();
+		break;
 	default:
 		usage();
 	} ARGEND;
