diff --git a/config.def.h b/config.def.h
index e06be66..de499a0 100644
--- a/config.def.h
+++ b/config.def.h
@@ -1,10 +1,10 @@
 /* See LICENSE file for copyright and license details. */
 
 /* interval between updates (in ms) */
-const unsigned int interval = 1000;
+const unsigned int interval = 1;
 
 /* text to show if no value can be retrieved */
-static const char unknown_str[] = "n/a";
+static const char unknown_str[] = "";
 
 /* maximum output string length */
 #define MAXLEN 2048
@@ -62,6 +62,21 @@ static const char unknown_str[] = "n/a";
  * wifi_essid          WiFi ESSID                      interface name (wlan0)
  */
 static const struct arg args[] = {
-	/* function format          argument */
-	{ datetime, "%s",           "%F %T" },
+	/* function       format        argument */
+	{ run_command,    "%s",         "echo ' '" },
+	{ run_command,    "%s",         "$HOME/bin/status/record" },
+	{ run_command,    "%s",         "$HOME/bin/status/tor" },
+	{ run_command,    "%s",         "$HOME/bin/status/vpn" },
+	{ run_command,    "%s",         "$HOME/bin/status/key-mon" },
+	{ run_command,    "%s",         "$HOME/bin/status/screenkey" },
+	{ run_command,    "%s ",        "$HOME/bin/status/network-wlan-tfc" },
+    { run_command,    "%s ",        "$HOME/bin/status/cpu-temp" },
+    { run_command,    "%s ",        "$HOME/bin/status/memory" },
+    { run_command,    "%s ",        "$HOME/bin/status/filesystem" },
+	{ run_command,    "%s ",        "$HOME/bin/status/volume" },
+	{ run_command,    "%s ",        "$HOME/bin/status/mailbox" },
+	{ run_command,    "%s ",        "$HOME/bin/status/datetime" },
+	{ run_command,    "%s ",        "$HOME/bin/status/bat-state" },
+	{ run_command,    "%s ",        "$HOME/bin/status/bat-capacity" },
+	{ run_command,    "%s",         "$HOME/bin/status/user" },
 };
