diff --git a/tabbed.c b/tabbed.c
index ff3ada0..c3826c2 100644
--- a/tabbed.c
+++ b/tabbed.c
@@ -317,6 +317,7 @@ drawbar(void)
 	XftColor *col;
 	int c, cc, fc, width;
 	char *name = NULL;
+	char tabtitle[256];
 
 	if (nclients == 0) {
 		dc.x = 0;
@@ -358,7 +359,9 @@ drawbar(void)
 		} else {
 			col = clients[c]->urgent ? dc.urg : dc.norm;
 		}
-		drawtext(clients[c]->name, col);
+		snprintf(tabtitle, sizeof(tabtitle), "%d: %s",
+		         c + 1, clients[c]->name);
+		drawtext(tabtitle, col);
 		dc.x += dc.w;
 		clients[c]->tabx = dc.x;
 	}
