diff --git a/config.mk b/config.mk
index ddf278a..4c4c5d5 100644
--- a/config.mk
+++ b/config.mk
@@ -30,7 +30,6 @@ STLDFLAGS = $(LIBS) $(LDFLAGS)
 #LIBS = -L$(X11LIB) -lm -lX11 -lutil -lXft \
 #       `$(PKG_CONFIG) --libs fontconfig` \
 #       `$(PKG_CONFIG) --libs freetype2`
-#MANPREFIX = ${PREFIX}/man
 
 # compiler and linker
 # CC = c99
diff --git a/st.1 b/st.1
index 39120b4..57ddfb8 100644
--- a/st.1
+++ b/st.1
@@ -22,6 +22,8 @@ st \- simple terminal
 .IR line ]
 .RB [ \-w
 .IR windowid ]
+.RB [ \-z
+.IR fontsize ]
 .RB [[ \-e ]
 .IR command
 .RI [ arguments ...]]
@@ -44,6 +46,8 @@ st \- simple terminal
 .IR title ]
 .RB [ \-w
 .IR windowid ]
+.RB [ \-z
+.IR fontsize ]
 .RB \-l
 .IR line
 .RI [ stty_args ...]
@@ -91,6 +95,10 @@ defines the window title (default 'st').
 embeds st within the window identified by
 .I windowid
 .TP
+.BI \-z " fontsize"
+sets the default fontsize to
+.I fontsize
+.TP
 .BI \-l " line"
 use a tty
 .I line
diff --git a/x.c b/x.c
index 2a3bd38..575c1f5 100644
--- a/x.c
+++ b/x.c
@@ -4,6 +4,7 @@
 #include <limits.h>
 #include <locale.h>
 #include <signal.h>
+#include <stdlib.h>
 #include <sys/select.h>
 #include <time.h>
 #include <unistd.h>
@@ -1145,7 +1146,7 @@ xinit(int cols, int rows)
 		die("could not init fontconfig.\n");
 
 	usedfont = (opt_font == NULL)? font : opt_font;
-	xloadfonts(usedfont, 0);
+	xloadfonts(usedfont, defaultfontsize);
 
 	/* colors */
 	xw.cmap = XDefaultColormap(xw.dpy, xw.scr);
@@ -2071,6 +2072,11 @@ main(int argc, char *argv[])
 	case 'v':
 		die("%s " VERSION "\n", argv0);
 		break;
+	case 'z':
+		defaultfontsize = strtod(EARGF(usage()), NULL);
+		if (!(defaultfontsize > 0))
+			usage();
+		break;
 	default:
 		usage();
 	} ARGEND;
