diff --git a/config.mk b/config.mk
index beafc35..0cbb002 100644
--- a/config.mk
+++ b/config.mk
@@ -1,5 +1,5 @@
 # st version
-VERSION = 0.8.3
+VERSION = 0.8.2
 
 # Customize below to fit your system
 
diff --git a/x.c b/x.c
index e5f1737..22c7dfc 100644
--- a/x.c
+++ b/x.c
@@ -1382,7 +1382,7 @@ xdrawglyphfontspecs(const XftGlyphFontSpec *specs, Glyph base, int len, int x, i
 
 	/* Change basic system colors [0-7] to bright system colors [8-15] */
 	if ((base.mode & ATTR_BOLD_FAINT) == ATTR_BOLD && BETWEEN(base.fg, 0, 7))
-		fg = &dc.col[base.fg + 8];
+		fg = &dc.col[base.fg];
 
 	if (IS_SET(MODE_REVERSE)) {
 		if (fg == &dc.col[defaultfg]) {
